### PR TITLE
add directory-listing plugin

### PR DESCRIPTION
**Pull request by artem-fedorov**

### CODE REVIEW

File: plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+    private Iterator<File> iterator;
+    private List<File> list;
+
+    @Override
+    public void iterationStart(LoopIterationEvent loopIterationEvent) {
+        // TODO: independentListPerThread
+
+        try {
+            if (list == null) {
```
Comment (undera) [inline plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java]: you can put this `if` into one below

File: plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                boolean isReRead = getReReadDirectoryOnTheEndOfList();
+                boolean isRewindOnTheEnd = getRewindOnTheEnd();
+
+                if (isReRead && isRewindOnTheEnd) {
+                    this.list = getDirectoryListing();
+                    this.iterator = list.iterator();
+                } else if (isRewindOnTheEnd) {
+                    this.iterator = list.iterator();
+                } else {
+                    return;
```
Comment (undera) [inline plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java]: What's this situation? Why we return without any action?

Comment (artem-fedorov) [inline plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java]: If list iterator at the end && (isRewindOnTheEnd == false) 

What we doing in this situation?

File: plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java (hunk 1 above)
Comment (artem-fedorov) [inline plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java]: no, next `if` for situation when iterator at the end of list

File: plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java (hunk 2 above)
Comment (undera) [inline plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java]: Stopping thread.

File: plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+import org.apache.jmeter.threads.JMeterContextService;
+import org.apache.jmeter.threads.JMeterVariables;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.*;
+
+public class DirectoryListing extends ConfigTestElement implements LoopIterationListener {
```
Comment (undera) [inline plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java]: Name it DirectoryListingConfig, that would explain the class better

File: plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                    this.iterator = list.iterator();
+                } else if (isRewindOnTheEnd) {
+                    this.iterator = list.iterator();
+                } else {
+                    // if the end of list && !isRewindOnTheEnd
+                    return;
+                }
+            }
+
+        } catch (FileNotFoundException ex) {
```
Comment (undera) [inline plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java]: Do you really need to catch this exception?

Comment (artem-fedorov) [inline plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryListing.java]: yes, here we must catch this exception and logging it or throw RuntimeException (failed all test)

File: plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryIterator.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+import org.apache.jmeter.threads.JMeterContextService;
+
+import java.io.File;
+import java.util.Iterator;
+import java.util.List;
+
+public class DirectoryIterator {
+
+    private final DirectoryListingConfig config;
```
Comment (undera) [inline plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryIterator.java]: bad decision. business logic class should not depend on its clients. It's client class that should depend on business logic class.
The solution is to make DirectoryIterator to hold all the flags as fields. Or have dedicated Configuration class. Or have bitmasks to not have too many methods and fields.

File: plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryIterator.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    public DirectoryIterator(DirectoryListingConfig config) {
+        this.config = config;
+        list = config.getDirectoryListing();
```
Comment (undera) [inline plugins/directory-listing/src/main/java/com/blazemeter/jmeter/DirectoryIterator.java]: No, it's this class that has to do directory listing.

File: site/dat/wiki/DirectoryListing.wiki [hunk 7]
```diff
@@ -0,0 +1,16 @@
+= Directory Listing Config =
+
+<span class=''>[/?search=jpgc-directory-listing <i class='fa fa-download'></i> Download]</span>
+
+Directory listing config plugin offers opportunities that allows to set file path from source directory into jmeter-variable in iteration loop.
```
Comment (undera) [inline site/dat/wiki/DirectoryListing.wiki]: opportunities vs capabilities

File: site/dat/wiki/DirectoryListing.wiki [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+This plugin have some config options, that influence for the behavior:
+  * "Use full path" - is this flag does not selected plugin will be set sub-path from source path into destination variable;
+  * "Random order"  - shuffle listing in random order;
+  * "Recursive listing" - meaning, that plugin will be looking for all sub-folders in source directory;
+  * "Rewind on end of list" - if the flag is selected and an iteration loop has reached the end, the new loop will be started;
+  * "Re-read directory on the end of list" - re-read directory before starting new iteration loop;
+  * "Independent list per thread" - determines that listing will be shared for all threads or each thread will be have own local copy.
```
Comment (undera) [inline site/dat/wiki/DirectoryListing.wiki]: "will have" or "will be having", never "will be have"

**Pull request merged**